### PR TITLE
Fix data loss when renaming custom editors

### PR DESCRIPTION
**Pull request by lramos15**
This PR is part of https://github.com/microsoft/vscode/issues/101370.

I don't love the solution but I believe the only perfect solution would be editor input and model reuse upon rename which would touch too many pieces for the amount of iteration left (I'm also not sure if I'm familiar enough with everything just yet to tackle that). So instead what we do here is trigger a backup on all renamed models and then notify new editor inputs that they were just renamed ones. These editor inputs then search for a backup and if there is one they restore it. There is visual sluggishness during the restore, but it's better than data loss.

### CODE REVIEW

File: src/vs/workbench/api/browser/mainThreadCustomEditors.ts [hunk 1]
```diff
@@ -138,9 +143,19 @@ export class MainThreadCustomEditors extends Disposable implements extHostProtoc
...
 				webviewInput.webview.extension = extension;
 
+				// If there's an old resource this was a move and we must resolve the backup at the same time as the webview
+				// This is because the backup must be ready upon model creation, and the input resolve method comes after
+				let backupID = webviewInput.backupId;
+				if (webviewInput.oldResource && !webviewInput.backupId) {
+					const backupIdentifier = { resource: webviewInput.oldResource, typeId: NO_TYPE_ID };
+					const backup = await this.workingCopyBackupService.resolve<CustomDocumentBackupData>(backupIdentifier);
+					backupID = backup?.meta?.backupId;
+					this.workingCopyBackupService.discardBackup(backupIdentifier);
```
Comment (bpasero) [inline src/vs/workbench/api/browser/mainThreadCustomEditors.ts:153]: This operation is `async`

File: src/vs/workbench/api/browser/mainThreadCustomEditors.ts [hunk 2]
```diff
@@ -138,9 +143,19 @@ export class MainThreadCustomEditors extends Disposable implements extHostProtoc
...
+				}
+
 				let modelRef: IReference<ICustomEditorModel>;
 				try {
-					modelRef = await this.getOrCreateCustomEditorModel(modelType, resource, viewType, { backupId: webviewInput.backupId }, cancellation);
+					modelRef = await this.getOrCreateCustomEditorModel(modelType, resource, viewType, { backupId: backupID }, cancellation);
```
Comment (bpasero) [inline src/vs/workbench/api/browser/mainThreadCustomEditors.ts:158]: Why not call it `backupId` and thus be able to remove `: backupID` entirely.

File: src/vs/workbench/api/browser/mainThreadCustomEditors.ts [hunk 3]
```diff
@@ -253,6 +268,31 @@ export class MainThreadCustomEditors extends Disposable implements extHostProtoc
...
+		}
+		e.waitUntil((async () => {
+			const models = [];
+			for (const file of e.files) {
+				if (file.source) {
+					models.push(...(await this._customEditorService.models.getAllModels(file.source)));
+				}
+			}
+			for (const model of models) {
+				if (model instanceof MainThreadCustomEditorModel) {
```
Comment (bpasero) [inline src/vs/workbench/api/browser/mainThreadCustomEditors.ts:285]: Why are all models backed up, shouldn't it only be the dirty ones?

**Review by bpasero — CHANGES_REQUESTED** [review]
Rather than literally using the working copy backup service and producing backups, can we not just do this:
* participate in move (as you do now)
* for each dirty custom model, call `backup`
* keep the backups around in some map (but do not store them via backup service)
* when the move operation has finished, restore all custom editors passing in the backup?

File: src/vs/workbench/api/browser/mainThreadCustomEditors.ts (hunk 1 above)
Comment (lramos15) [reply src/vs/workbench/api/browser/mainThreadCustomEditors.ts:153]: Yeah, but I don't need to await it as it doesn't matter when the backup gets discarded. At least that was my thought.

**Comment by lramos15** [comment]
I believe your suggestion of a map would work. I didn't know if having the working copy backup service the source of truth for backups was important.

**Comment by bpasero** [comment]
> I believe your suggestion of a map would work. I didn't know if having the working copy backup service the source of truth for backups was important.

No, not needed, as long as you do not need the backup to survive a restart.

**Comment by bpasero** [comment]
Slightly weird for extension though: it possibly gets a backup passed in from a different file. Not sure if that is a problem.

**Comment by lramos15** [comment]
> Slightly weird for extension though: it possibly gets a backup passed in from a different file. Not sure if that is a problem.

The way I implemented the hex editor (and the editor sample) is that you load the data from backup but the actual URI you save to is different. Therefore it's ok that the backup has a separate URI.

**Pull request merged**